### ISSUE TITLE
ulimits not propagated 

### ISSUE THREAD

**Issue by ryanday36** [flux-framework/flux-core#4743]
We were getting reports of unexpected seg faults while compiling in a Flux instance, and found that the stack size was being limited in the flux instance:
```
[day36@rzwhippet17:~]$ ulimit -s 
unlimited
[day36@rzwhippet17:~]$ flux mini alloc -N1 --exclusive
[day36@rzwhippet2:~]$ ulimit -s
8192
[day36@rzwhippet2:~]$
```
Is there something that we should be configuring to get ulimits propagated into the Flux instance?

**Comment by grondo** [comment]
We don't currently support propagating ulimits, so resource limits are being inherited from the brokers.
Until we can implement this, a more reasonable stack size could be set via `/etc/security/limits.conf` for logins and Flux jobs if we enable PAM config for the IMP (See [flux-config-security-imp(5)](https://flux-framework.readthedocs.io/projects/flux-security/en/latest/man5/flux-config-security-imp.html#flux-config-security-imp-5).

**Comment by grondo** [comment]
Implementation of resource limit propagation could be a little tricky. It would be trvial to add an `atttributes.system.shell.options.rlimits` section to jobspec, and have a simple job shell plugin that would attempt to set any limits defined therein. However, the job shell doesn't run with any elevated privileges, so it would be limited by all current hard limits. Therefore, we'd also have to configure PAM for the IMP as suggested above to increase the hard limits to something reasonable for user jobs.

If that sounds like a reasonable first approach, I could probably propose something soon.

**Comment by garlick** [comment]
What does it mean to propagate ulimits?  Capture the user's limits in the jobspec at submit time, and then in the job
- set initial limits with the IMP PAM config as mentioned above
- have a shell plugin attempt to apply the jobspec limits

I get why that would be desirable for jobs submitted from a batch script because then you could put `ulimit` commands in the batch script, but would it be a bad thing to propagate someone's login node settings when the batch job is submitted (they might set an unlimited core during development, but not want thousands of large core files in the job)?

**Comment by grondo** [comment]
I think this is another "we're used to what Slurm does, so everything else should behave the same" issue. Slurm propagates all current rlimits by default to all jobs, unless the user specifies otherwise via the `--propagate=LIST` option. (At least that is how I remember it working).

**Comment by ryanday36** [comment]
Thanks @grondo and @garlick! I've asked Clay to look at enabling the IMP PAM configs and setting the limits to match what users get by default. I think that'll go a long way toward fixing this.

In general, I think it's best for the environment/limits/ etc in the job shell to match the environment/limits/etc in the user's shell. Then, if the user wants something different in their job, they're making changes from a baseline that they know (or at least can know without running a test job). Also, a lot of users have no idea about any of this stuff, so don't even know where to start when things behave differently in their job than they did on the login node.

**Comment by chu11** [comment]
We've had conversations in the past about limits being configurable / propagated via systemd job-exec.  But I suppose that's too long out?

**Comment by grondo** [comment]
I vaguely remember that, but now that I think about it, I'm not sure how that helps here. It will be trivial to add a shell plugin to adjust limits within any hard limits imposed by PAM or the system, and systemd couldn't adjust those hard limits either since it is running as the `flux` user, which has no privileges. I think where sdexec is required is for setting up cgroups via `AllowedCPUs` etc...

**Comment by grondo** [comment]
We might also need a way to make the set of limits automatically propagated configurable. In order to allow jobs to have unlimited max locked memory on compute but not login nodes, we have
```
PropagateResourceLimitsExcept=MEMLOCK
```
in our `slurm.conf`. This allows the soft limit set by `/etc/security/limits.conf` to persist and not be overridden by the job's propagated limits.

We'll need to allow a similar outcome in Flux.

**Comment by grondo** [comment]
Should have been closed by #4745
